scripts/subspace_mlp_demo.py

```diff
@@ -0,0 +1,191 @@
+from functools import partial
+
+
+def get_datasets():
+    """
+    Load MNIST train and test datasets into memory
+    """
+    ds_builder = tfds.builder("mnist")
+    ds_builder.download_and_prepare()
+    train_ds = tfds.as_numpy(ds_builder.as_dataset(split="train", batch_size=-1))
+    test_ds = tfds.as_numpy(ds_builder.as_dataset(split="test", batch_size=-1))
+    train_ds["image"] = jnp.float32(train_ds["image"]) / 255.
+    test_ds["image"] = jnp.float32(test_ds["image"]) / 255.
+    return train_ds, test_ds
+
+
+class MLP(nn.Module):
+    @nn.compact
+    def __call__(self, x):
+        x = nn.relu(nn.Dense(784)(x))
+        x = nn.relu(nn.Dense(200)(x))
+        x = nn.relu(nn.Dense(200)(x))
+        x = nn.Dense(10)(x)
+        return nn.log_softmax(x)
+
+
+@jax.vmap
+def cross_entropy_loss(logits, label):
+    return -logits[label]
+
+
+@jax.jit
+def normal_loss(params, batch):
+    logits = MLP().apply({"params": params}, batch["image"])
+    logits = jax.nn.log_softmax(logits)
+    loss = jnp.mean(cross_entropy_loss(logits, batch["label"]))
+    return loss
+
+
+@jax.jit
+def normal_accuracy(params,batch):
+    logits = MLP().apply({"params": params}, batch["image"])
+    logits = jax.nn.log_softmax(logits)
+    return jnp.mean(jnp.argmax(logits, -1) == batch["label"])
+
+
+@jax.jit
+def theta_to_flat_params(theta,M,flat_params0):
+    return jnp.matmul(theta, M)[0] + flat_params0
+
+
+def projected_loss(theta_subspace, batch, M, flat_params0, reconstruct_fn):
+    """
+    Project the subspace model weights onto the full model weights and
+    compute the loss.
+            w(theta) = w_init + A * theta
+    1. Project theta_subspace ∈ R^d => theta ∈ R^D
+    2. Compute loss of the model w.r.t. theta_subspace
+    """
+    projected_params = theta_to_flat_params(theta_subspace, M, flat_params0)
+    projected_params = reconstruct_fn(projected_params)
+    return normal_loss(projected_params, batch)
+
+
+@jax.jit
+def adam_update(grads, params, mass, velocity, hyperparams):
+    mass = hyperparams["beta_1"] * mass + (1.0 - hyperparams["beta_1"]) * grads
+    velocity = hyperparams["beta_2"] * velocity + (1.0 - hyperparams["beta_2"]) * (grads ** 2.0)
+    # Bias correction
+    hat_mass = mass / (1 - hyperparams["beta_1"])
+    hat_velocity = velocity / (1 - hyperparams["beta_2"])
+    # Update
+    params = params - hyperparams["lr"] / (jnp.sqrt(hat_velocity) + hyperparams["epsilon"]) * hat_mass
+    return params, mass, velocity
+
+
+def generate_projection(key, d, D):
+    M = random.normal(key, shape=(d,D))
+    M = M / jnp.linalg.norm(M, axis=-1, keepdims=True)
+    return M
+
+
+def subspace_learning(key, model, datasets, d, hyperparams, n_epochs=300):
+    key_params, key_subspace = random.split(key)
+    _, num_features = train_ds["image"].shape
+
+    x0 = jnp.zeros(num_features)
+    w_init = model().init(key_params, x0)["params"]
+    w_init_flat, reconstruct_fn = jax.flatten_util.ravel_pytree(w_init)
```

Comment:
rename `reconstruct_fn` to `unflatten_fn`?

Reply:
Renaming to `flat_to_pytree_fn`